### ISSUE TITLE
Notification.permission always shows "granted"

### ISSUE THREAD

**Issue by pfrazee** [electron/electron#11221]
* Electron version: 1.8.2-beta
* Operating system: MacOS 10.12

### Expected behavior

`Notification.permission` should reflect whether permission has been granted to the `webContents` to show notifications.

### Actual behavior

`Notification.permission` is always set to "granted."

### How to reproduce

https://github.com/pfrazee/electron-notifications-permission

```
git clone https://github.com/pfrazee/electron-notifications-permission.git
cd electron-quick-start
npm install
npm start
```

You'll find in the render thread's devtools that permission is 'granted' even though a permission handler is denying permission.

Permission handler: https://github.com/pfrazee/electron-notifications-permission/blob/master/main.js#L16

Render thread: https://github.com/pfrazee/electron-notifications-permission/blob/master/renderer.js

**Comment by ckerr** [comment]
Using pfrazee's test case I'm seeing the same behavior on Linux.

**Comment by MarshallOfSound** [comment]
This afaik has always been the case, I'll run through the current situation and my recommendation at the end.

There is a semantic difference in the Notification spec that `Notification.permision` is not the same as the result of `Notification.requestPermission`.  `Notification.permission` is a synchronous getter that must return `granted`, `denied` or `default` for the origin of the "request".

The `requestPermission` method should should check the value of `permission` and do different things in each case.

* `granted`
  * Resolve the promise immediately with "granted"
* `denied`
  * Resolve the promise immediately with "denied"
* `default`
  * Ask the user (in this case our Electron permissions API) whether showing a notification is allowed
  * Update the `permission` property appropriately
  * Resolve the promise the the new value of `permission`

Currently we do not adhere to this spec, first off our [default setting for `permission`](https://github.com/electron/electron/blob/master/brightray/browser/platform_notification_service.cc#L100) is "granted".  This should be "default" (in chromium land this is "ask") which indicates to the web app that it needs to ask for permission.

Once a permission request has been initiated inside Electron (explicitly or implicitly) we should update the `permission` property with the last responded value.

Here is where it gets complicated, the determined `granted` / `denied` state in a normal spec compliant browser would persist between page loads.  In Electron has the permission request handler could in theory "change it's mind" between page loads so we're probably going to need to add another API (`webContents.setSynchronousNotificationPermisssionHandler()` or something similar) that can synchronously respond whether or not notifications have been explicitly "granted" or "denied" in the past.  This information should be persisted / maintained by Electron app developers not by Electron core.

TLDR:
* Change default of `Notification.permission` to `default`
* Update the value of `Notification.permission` on each request if no getter has been provided
* If a synchronous getter has been provided map `Notification.permission` to that value

Happy to work on these changes if we're OK with the TLDR bit at the bottom 👍

**Comment by MarshallOfSound** [comment]
As an addition although this change would IMO make us spec compliant, it would be a breaking API change (what used to always be `granted` will now be `default` and will cause different behavior in certain sites.

**Comment by sofianguy** [comment]
Thank you for taking the time to report this issue and helping to make Electron better.

The version of Electron you reported this on has been superseded by newer releases. 

If you're still experiencing this issue in Electron v2.0.x or later, please add a comment specifying the version you're testing with and any other new information that a maintainer trying to reproduce the issue should know.

I'm setting the `blocked/need-info` label for the above reasons.

Thanks in advance! Your help is appreciated.

**Comment by sofianguy** [comment]
Thank you for your issue!

We haven't gotten a response to our questions in our comment above. With only the information that is currently in the issue, we don't have enough information to take action. I'm going to close this but don't hesitate to reach out if you have or find the answers we need, we'll be happy to reopen the issue.

**Comment by ionalexandrusecara** [comment]
Hey everyone!

I would like to re-open this issue. Following the documentation at [https://www.electronjs.org/docs/tutorial/notifications](https://www.electronjs.org/docs/tutorial/notifications), `Notification.permission` always returns granted, except the first time the app is opened/installed and either a notification or a request for permission is issued - in which cause, default is returned. 

It would be useful to have a way to tell whether the notification permission is denied, so we can prompt users to change their settings.

**Comment by stefvw93** [comment]
This should be reopened. I am using 15.3.0 and have the exact same issue as @ionalexandrusecara . It would be awesome to properly read permission state and update the application UI accordingly.

**Comment by mqliutie** [comment]
This issue should be reopened. I have to get the permission in order to tell user to allow notification

**Comment by annie-elequin** [comment]
@sofianguy This issue should be reopened. I experience the same behavior as @ionalexandrusecara , on MacOS 12.5 and Electron version 20.0.0, both `Notificaiton.permission` and `Notification.requestPermission()` return "granted" even if I deny notifications in System Settings.

**Comment by danline** [comment]
I am experiencing the same issue. Notification.permission is always 'granted' although I am returning `false` in 
`session.setPermissionCheckHandler`

Electron version: 19.0.1

Would be great if there is a solution or workaround to this issue!

**Comment by Alireza17224** [comment]
Same issue here

**Comment by M-ZubairAhmed** [comment]
Same issue with Mattermost as well, please consider reopening this issue for possible solution change to the Notification spec in electron.
